src/main/java/org/jabref/gui/citationkeypattern/GenerateCitationKeyAction.java

```diff
@@ -90,9 +102,12 @@ private void generateKeys() {
             final NamedCompound compound = new NamedCompound(Localization.lang("Autogenerate citation keys"));
             CitationKeyGenerator keyGenerator =
                     new CitationKeyGenerator(databaseContext, Globals.prefs.getCitationKeyPatternPreferences());
+            int entriesDone = 0;
             for (BibEntry entry : entries) {
                 keyGenerator.generateAndSetKey(entry)
                             .ifPresent(fieldChange -> compound.addEdit(new UndoableKeyChange(fieldChange)));
+                entriesDone++;
+                backgroundProgressProperty.setValue(new BackgroundTask.BackgroundProgress(entriesDone, entries.size()));
```

Comment:
I think this will lead to problem since you are updating a property that is eventually displayed in the ui from a background task. This is not allowed in JavaFX. 

Instead one should create a class extending `BackgroundTask` and use the `updateProgress` method, which takes care of that the events are propagated in the right thread. See https://github.com/JabRef/jabref/blob/bb011c9313367a28990ae213b3920fe6cd10d1dc/src/main/java/org/jabref/gui/externalfiles/FileDownloadTask.java#L30 for an example.

Reply:
Yep, a good example is in ImportHandler. Needs to be wrapped on the FX thread though
```
    DefaultTaskExecutor.runInJavaFXThread(() -> {
                        updateMessage(Localization.lang("Processing file %0", file.getFileName()));
                        updateProgress(counter, files.size() - 1);
                    });
```

Reply:
I tried to avoid introducing a new Task for each BackgroundTask that is to be displayed in the GUI.
What do you think of introducing a _trackProgress_ method in BackgroundTask that takes a BackgroundProgress property and Binds it to updateProgress? Would that solve that issue? If so, would it be preferable or do you prefer having an additional class?

Reply:
I think it's a good idea to simplify the creation of tasks that track progress.

> What do you think of introducing a trackProgress method in BackgroundTask that takes a BackgroundProgress property and Binds it to updateProgress?

So the code would somewhat like the following?
```java
Property<...> progress = new ...
task = BackgroundTask.wrap(() => { do stuff; progress.set(100) }).trackProgress(progress)
```
This can work, but you still need to handle all the thread synchronization in the trackProgress method.

Alternatively, one could pass all the update* methods to the `wrap` function:
```java
task = BackgroundTask.wrap((updateProgress, updateMessage) => { do stuff; updateProgress(100) })
```

But to be honest, I'm not sure if it's worth all the trouble. It's not much shorter/clearer than using anonymous classes:
```java
task = new BackgroundTask<V>() {
            @Override
            protected V call() throws Exception {
                do stuff;
                updateProgress(100)
            }
        };
```

(But I agree that most instances of our background tasks can be converted to these anonymous classes instead of defining them in a new file).

Reply:
Anonymous classes seem to be the consensus. Let's do it that way.